flink-core/src/main/java/org/apache/flink/api/connector/sink/Sink.java

```diff
@@ -46,44 +54,81 @@
+     * #createWriter(InitContext, List)} on recovery.
+     *
+     * @return the serializer of the writer's state type.
+     */
+    Optional<SimpleVersionedSerializer<WriterStateT>> getWriterStateSerializer();
+
+    /**
+     * Creates a {@link Committer} which is part of a 2-phase-commit protocol. The {@link
+     * SinkWriter} creates committables through {@link SinkWriter#prepareCommit(boolean)} in the
+     * first phase. The committables are then passed to this committer and persisted with {@link
+     * Committer#commit(List)}. If a committer is returned, the sink must also return a {@link
+     * #getCommittableSerializer()}.
      *
-     * @return A committer.
-     * @throws IOException if fail to create a committer.
+     * @return A committer for the 2-phase-commit protocol.
+     * @throws IOException for any failure during creation.
      */
     Optional<Committer<CommT>> createCommitter() throws IOException;
 
     /**
-     * Creates a {@link GlobalCommitter}.
+     * Creates a {@link GlobalCommitter} which is part of a 2-phase-commit protocol. The {@link
+     * SinkWriter} creates committables through {@link SinkWriter#prepareCommit(boolean)} in the
+     * first phase. The committables are then passed to the Committer and persisted with {@link
+     * Committer#commit(List)} which also can return an aggregated committable. This aggregated
+     * committables are passed to this {@link GlobalCommitter} of which only a single instance
+     * exists. If a global committer is returned, the sink must also return a {@link
+     * #getCommittableSerializer()} and {@link #getGlobalCommittableSerializer()}.
      *
-     * @return A global committer.
-     * @throws IOException if fail to create a global committer.
+     * @return A global committer for the 2-phase-commit protocol.
+     * @throws IOException for any failure during creation.
      */
     Optional<GlobalCommitter<CommT, GlobalCommT>> createGlobalCommitter() throws IOException;
 
-    /** Returns the serializer of the committable type. */
+    /**
+     * Returns the serializer of the committable type. The serializer is required iff the sink has a
+     * {@link Committer} or {@link GlobalCommitter}.
+     */
     Optional<SimpleVersionedSerializer<CommT>> getCommittableSerializer();
 
-    /** Returns the serializer of the aggregated committable type. */
+    /**
+     * Returns the serializer of the aggregated committable type. The serializer is required iff the
+     * sink has a {@link GlobalCommitter}.
+     */
     Optional<SimpleVersionedSerializer<GlobalCommT>> getGlobalCommittableSerializer();
 
-    /** Return the serializer of the writer's state type. */
-    Optional<SimpleVersionedSerializer<WriterStateT>> getWriterStateSerializer();
+    /**
+     * A list of state names of sinks from which the state can be restored. For example, the new
+     * file sink can resume from the state of an old streaming file sink as a drop-in replacement
```

Comment:
I can't exactly use a JavaDoc link because of circular dependency issues. But I have used the proper class names in the code tag to make it more obvious.